sycl/doc/design/DeviceConfigFile.md

```diff
@@ -0,0 +1,297 @@
+
+### New `TableGen` backend
+Note: This [guide](https://llvm.org/docs/TableGen/BackGuide.html) details how
+to implement new TableGen backends. Also, the [Search
+Indexes](https://llvm.org/docs/TableGen/BackEnds.html#search-indexes) backend
+already does something very similar to what we seek. It generates a table that
+provides a lookup function, but it cannot be extended with new entries. We can
+use _Search Indexes_ backend as inspiration for ours. 
+
+Our backend should generate a map where the key is the target name and the value
+is an object of a custom class/struct including all the information required. 
+
+The original `.td` file should look like the one below: 
+``` 
+include "llvm/TableGen/SearchableTable.td"
+// AspectEnum, AspectEntry, and all the aspects definitions could be outlined
+// to another .td file that could be included into this file
+def AspectEnum : GenericEnum {
+  let FilterClass = "AspectEntry";
+  let NameField = "Name";
+  let ValueField = "Encoding";
+}
+
+class AspectEntry<bits<16> enc> {
+  string Name = NAME;
+  bits<16> Encoding = enc;
+}
+
+def AspectCpu : AspectEntry<1>;
+def AspectGpu : AspectEntry<2>;
+def AspectAccelerator : AspectEntry<3>;
+def AspectCustom : AspectEntry<4>;
+def AspectFp16 : AspectEntry<5>;
+def AspectFp64 : AspectEntry<6>;
+def AspectImage : AspectEntry<9>;
+def AspectOnline_compiler : AspectEntry<10>;
+def AspectOnline_linker : AspectEntry<11>;
+def AspectQueue_profiling : AspectEntry<12>;
+def AspectUsm_device_allocations : AspectEntry<13>;
+def AspectUsm_host_allocations : AspectEntry<14>;
+def AspectUsm_shared_allocations : AspectEntry<15>;
+def AspectUsm_system_allocations : AspectEntry<17>;
+def AspectExt_intel_pci_address : AspectEntry<18>;
+def AspectExt_intel_gpu_eu_count : AspectEntry<19>;
+def AspectExt_intel_gpu_eu_simd_width : AspectEntry<20>;
+def AspectExt_intel_gpu_slices : AspectEntry<21>;
+def AspectExt_intel_gpu_subslices_per_slice : AspectEntry<22>;
+def AspectExt_intel_gpu_eu_count_per_subslice : AspectEntry<23>;
+def AspectExt_intel_max_mem_bandwidth : AspectEntry<24>;
+def AspectExt_intel_mem_channel : AspectEntry<25>;
+def AspectUsm_atomic_host_allocations : AspectEntry<26>;
+def AspectUsm_atomic_shared_allocations : AspectEntry<27>;
+def AspectAtomic64 : AspectEntry<28>;
+def AspectExt_intel_device_info_uuid : AspectEntry<29>;
+def AspectExt_oneapi_srgb : AspectEntry<30>;
+def AspectExt_oneapi_native_assert : AspectEntry<31>;
+def AspectHost_debuggable : AspectEntry<32>;
+def AspectExt_intel_gpu_hw_threads_per_eu : AspectEntry<33>;
+def AspectExt_oneapi_cuda_async_barrier : AspectEntry<34>;
+def AspectExt_oneapi_bfloat16_math_functions : AspectEntry<35>;
+def AspectExt_intel_free_memory : AspectEntry<36>;
+def AspectExt_intel_device_id : AspectEntry<37>;
+def AspectExt_intel_memory_clock_rate : AspectEntry<38>;
+def AspectExt_intel_memory_bus_width : AspectEntry<39>;
+def AspectEmulated : AspectEntry<40>;
```

Comment:
Admittedly, this further intensifies my worries about using the aspect values in the original config file design. By adding this, we suddenly need to keep this non-runtime file in sync with the runtime, which we have previously tried our best to not need.

Would it be possible to make this completely oblivious to existing aspects and just require the users to provide the specific names or ID? The IR should contain a mapping of names and IDs, so either should be enough to work with in the end, at least for IR passes.

Reply:
> However, I think that actual numbers are also required for other uses of this config, such as: [diagnostic messages from -fsycl-fixed-targets](https://github.com/intel/llvm/blob/sycl/sycl/doc/design/OptionalDeviceFeatures.md#compiler-switch-to-diagnose-optional-feature-usage) or [defining a macro for any_device_has/all_devices_have](https://github.com/intel/llvm/blob/sycl/sycl/doc/design/DeviceAspectTraitDesign.md#changes-to-the-compiler-driver).

I don't think those strictly need the values. For the fixed-targets I would assume it would have access to the mappings either way, otherwise it would not be able to produce any useful diagnostics to the user about what aspects are problematic. For the `any_device_has`/`all_devices_have`, the macros they generate only use the value because that fit with the original design of the configuration file using values. We could change the macros to use the aspect names instead. It may even make the macros more readable, albeit more verbose.

Reply:
> Does the !sycl_aspects metadata get propagated through llvm-link, so that it is available to sycl-post-link?

Linker shouldn't drop the metadata. It may be altered a bit if several input modules contain the same metadata, i.e. they will be merged - we probably need to check if that would still be recoverable for our needs in `sycl-post-link`, but likely it will be.

>I intend to change the design for AOT optional features, so we no longer have the `sycl-aspect-filter` tool. Instead, `sycl-post-link` will read the config file and use that information to filter the kernels that are included in each device image.

BTW, @gmlueck, could you please share more details here? I though that the main motivation for `sycl-aspect-filter` tool is to overcome clang driver limitation of having statically known list of commands. Are we going to emit "reduced" modules for some of targets, i.e. remove kernels which are not supported there?

Also, to state this here explicitly: we had an offline sync some time ago and I don't have objections against using strings instead of integers as aspects identification.

Reply:
> BTW, @gmlueck, could you please share more details here? I though that the main motivation for `sycl-aspect-filter` tool is to overcome clang driver limitation of having statically known list of commands. Are we going to emit "reduced" modules for some of targets, i.e. remove kernels which are not supported there?

Yes, this was my thought.  The current POR is for `sycl-post-link` to emit all the kernels and then filter out the incompatible ones with a separate tool `sycl-aspect-filter`.  Rather than doing this, I think `sycl-post-link` can read the config file and filter out the incompatible kernels.  My [proposed design](https://github.com/intel/llvm/pull/9127) for "if_device_has" requires the `sycl-post-link` tool to read the config file anyways, so it may as well also do the filtering.

I think the list of commands is still known statically because the set of AOT targets is known statically (from the `-fsycl-targets` option).  For example, if the user compiles for two Intel GPU AOT targets with `-fsycl-targets=intel_gpu_pvc,intel_gpu_dg1`, then the compiler driver can invoke `sycl-post-link` like:

```
sycl-post-link [....] -o intel_gpu_pvc=<table1> -o intel_gpu_dg1=<table2>
```

In this case file table `<table1>` will contain the device images for PVC and `<table2>` will contain the device images for DG1.  Each file table will then be processed, as normal, through `llvm-for-each` to convert LLVM bitcode to SPIR-V and then compiling the SPIR-V via `ocloc`.